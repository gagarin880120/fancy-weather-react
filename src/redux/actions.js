import convertCodeToCountryName from '../utils/country-codes';

function setLanguage(language) {
  return {
    type: 'LANGUAGE',
    language,
  };
}

function setAddress(address) {
  return {
    type: 'ADDRESS',
    address,
  };
}

function setLocation(latitude, longitude) {
  return {
    type: 'LOCATION',
    latitude,
    longitude,
  };
}

function setCurrentDate(currentDate) {
  return {
    type: 'CURRENT_DATE',
    currentDate,
  };
}

function setCurrentDateInterval(currentDateInterval) {
  return {
    type: 'CURRENT_DATE_INTERVAL',
    currentDateInterval,
  };
}

function setCurrentWeather(weatherObj) {
  return {
    type: 'CURRENT_WEATHER',
    weatherObj,
  };
}

function setWeeklyWeather(weeklyWeatherArr) {
  return {
    type: 'WEEKLY_WEATHER',
    weeklyWeatherArr,
  };
}

function getCurrentDate(lat, lng) {
  const API_KEY = 'ST1WDEJRNQDM';
  return (dispatch) => fetch(`http://api.timezonedb.com/v2.1/get-time-zone?key=${API_KEY}`
  + `&format=json&by=position&lat=${lat}&lng=${lng}`)
    .then((res) => res.json())
    .then((data) => {
      dispatch(setCurrentDate(data.formatted));
    })
    .catch((e) => console.log(e));
}

function getWeather(lat, lon) {
  const API_KEY = 'cf236ef1f2e94b658f9c745520d1fe1a';
  return (dispatch) => fetch(`https://api.weatherbit.io/v2.0/forecast/daily?&lat=${lat}&lon=${lon}&days=8&lang=en&key=${API_KEY}`)
    .then((res) => res.json())
    .then((data) => {
      dispatch(setCurrentWeather(data.data[0]));
      dispatch(setWeeklyWeather(data.data.slice(1)));
    })
    .catch((e) => console.log(e));
}

function getDefaultAddress() {
  const API_KEY = '60e8ad51f55486';
  return (dispatch) => fetch(`https://ipinfo.io?token=${API_KEY}`)
    .then((res) => res.json())
    .then((data) => {
      dispatch(setAddress(`${data.city}, ${convertCodeToCountryName(data.country, 'en')}`));
      const lat = data.loc.split(',')[0];
      const lon = data.loc.split(',')[1];
      dispatch(setLocation(lat, lon));
      dispatch(getWeather(lat, lon));
      dispatch(setCurrentDateInterval(setInterval(() => dispatch(getCurrentDate(lat, lon)), 1000)));
    })
    .catch((e) => console.log(e));
}

export {
  setAddress, setCurrentWeather, getDefaultAddress, setLanguage,
};
