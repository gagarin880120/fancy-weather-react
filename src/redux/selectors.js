import convertCodeToName from '../utils/openweatherCodes';

const selectLanguage = (state) => state.language;
const selectAddress = (state) => state.address;

const selectCurrentDate = (state) => {
  if (state.currentDate) {
    const options = {
      weekday: 'short', day: 'numeric', month: 'long', hour: 'numeric', minute: 'numeric',
    };
    return new Date(state.currentDate).toLocaleDateString('en-GB', options);
  }
  return null;
};

const selectWeeklyWeather = (state) => {
  if (state.weeklyWeather) {
    return state.weeklyWeather.map((v) => ({
      ...v,
      wind_spd: Math.round(v.wind_spd),
      temp: Math.round(v.temp),
      weekday: new Date(v.datetime).toLocaleDateString('en-GB', { weekday: 'long' }),
      icon: convertCodeToName(
        v.weather.icon.slice(-1) + v.weather.code,
      ),
    }));
  }
  return null;
};

const selectCurrentWeather = (state) => {
  if (state.currentWeather) {
    return {
      ...state.currentWeather,
      wind_spd: Math.round(state.currentWeather.wind_spd),
      temp: Math.round(state.currentWeather.temp),
      icon: convertCodeToName(
        state.currentWeather.weather.icon.slice(-1) + state.currentWeather.weather.code,
      ),
    };
  }
  return null;
};

const selectLatitude = (state) => state.latitude;
const selectLongitude = (state) => state.longitude;
const selectBackgroundImageURL = (state) => state.backgroundImageURL;
const selectCurrentDateInterval = (state) => state.currentDateInterval;
const selectCountryFlagURL = (state) => state.countryFlagURL;

export {
  selectAddress, selectCurrentWeather, selectLanguage, selectLatitude, selectLongitude,
  selectCurrentDate, selectWeeklyWeather, selectBackgroundImageURL, selectCurrentDateInterval,
  selectCountryFlagURL,
};
