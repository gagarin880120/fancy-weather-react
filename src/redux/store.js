import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  language: 'en',
  latitude: '',
  longitude: '',
  address: '',
  currentDate: '',
  currentDateInterval: null,
  currentWeather: null,
  weeklyWeather: null,
  backgroundImageURL: '',
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'LANGUAGE':
      return { ...state, language: action.language };
    case 'LOCATION':
      return { ...state, latitude: action.latitude, longitude: action.longitude };
    case 'ADDRESS':
      return { ...state, address: action.address };
    case 'CURRENT_DATE':
      return { ...state, currentDate: action.currentDate };
    case 'CURRENT_DATE_INTERVAL':
      return { ...state, currentDateInterval: action.currentDateInterval };
    case 'CURRENT_WEATHER':
      return { ...state, currentWeather: action.weatherObj };
    case 'WEEKLY_WEATHER':
      return { ...state, weeklyWeather: action.weeklyWeatherArr };
    case 'BACKGROUND_IMAGE_URL':
      return { ...state, backgroundImageURL: action.imageURL };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
