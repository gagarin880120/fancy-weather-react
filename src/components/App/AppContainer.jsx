import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import App from './App';
import { selectBackgroundImageURL } from '../../redux/selectors';

export function AppContainer({ backgroundImageURL }) {
  console.log(backgroundImageURL);
  return (
    <App
      backgroundImageURL={backgroundImageURL}
    />
  );
}

export const mapStateToProps = (state) => ({
  backgroundImageURL: selectBackgroundImageURL(state),
});

AppContainer.propTypes = {
  backgroundImageURL: PropTypes.string,
};

AppContainer.defaultProps = {
  backgroundImageURL: '',
};

export default connect(mapStateToProps)(AppContainer);
