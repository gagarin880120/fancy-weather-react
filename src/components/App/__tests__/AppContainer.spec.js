import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { AppContainer, mapStateToProps, mapDispatchToProps } from '../AppContainer';
import { getDefaultAddress } from '../../../redux/actions';

jest.mock('mapbox-gl', () => ({
  mapboxgl: null,
}));

jest.mock('../../../redux/actions', () => ({
  getDefaultAddress: jest.fn()
}))

describe('AppContainer', () => {
  describe('component', () => {
    test('should render without crashing', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<AppContainer />);

      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    test('should return right value', () => {
      const initialState = {
        language: 'en',
        temperatureScale: 'celsius',
        backgroundImageURL: 'url',
        isModalOpen: false,
        currentWeather: {
          rh: 15,
          temp: 12,
          wind_spd: 3,
          weather: {
            code: '801',
            icon: 'c02d'
          }
        },
        currentDate: '2020-07-20 15:09:14',
        weeklyWeather: [
          {
            datetime: '2020-07-21',
            temp: 19.9,
            weather: {
              code: 804,
              icon: 'c04d',
            }
          },
          {
            datetime: '2020-07-22',
            temp: 15.6,
            weather: {
              code: 804,
              icon: 'c04d',
            }
          },
          {
            datetime: '2020-07-23',
            temp: 15.3,
            weather: {
              code: 804,
              icon: 'c04d',
            }
          }
        ],
      }

      expect(mapStateToProps(initialState)).toEqual({
        backgroundImageURL: 'url',
        isModalOpen: false,
        currentWeather: {
          rh: 'Humidity: 15%',
          wind_spd: 'Wind: 3m/s',
          temp: 12,
          icon: 'DayCloudyGusts',
          weather: {
            code: '801',
            icon: 'c02d',
          },
        },
        currentDate: 'Mon, 20 July, 15:09',
        weeklyWeather: [{
          datetime: '2020-07-21',
          weekday: 'Tuesday',
          temp: 20,
          icon: 'DaySunnyOvercast',
          weather: {
            code: 804,
            icon: 'c04d',
          }
        },
        {
          datetime: '2020-07-22',
          weekday: 'Wednesday',
          temp: 16,
          icon: 'DaySunnyOvercast',
          weather: {
            code: 804,
            icon: 'c04d',
          }
        },
        {
          datetime: '2020-07-23',
          weekday: 'Thursday',
          temp: 15,
          icon: 'DaySunnyOvercast',
          weather: {
            code: 804,
            icon: 'c04d',
          }
        }],
      })
    });
  });

  describe('mapDispatchToProps', () => {
    afterEach(() => {
      jest.clearAllMocks();
    })

    test('onAppLoad should dispatch getDefaultAddress action', () => {
      const dispatch = jest.fn();
      const {onAppLoad} = mapDispatchToProps(dispatch);

      onAppLoad();

      expect(getDefaultAddress).toHaveBeenCalled();
    });
  });
});
