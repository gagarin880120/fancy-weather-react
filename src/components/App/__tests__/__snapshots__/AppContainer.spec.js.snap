// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer component should render without crashing 1`] = `
<App
  backgroundImageURL=""
  isModalOpen={false}
/>
`;
